Support custom colors for icons

Mega-Volti opened the pull request:
## Proposed change

<!--
Please include a summary of the change. Screenshots and / or videos can also be helpful if appropriate.

*** Please see the development guidelines for new widgets: https://gethomepage.dev/en/more/development/#service-widget-guidelines
*** If you do not follow these guidelines your PR will likely be closed without review.

New service widgets should include example(s) of relevant relevant API output as well as a PR to the docs for the new widget.
-->

Enables appending a color code (e.g. "#123456") to all mdi and si icons, in order to change their color to a per-icon custom one.
If no color code is appended, defaults to current behaviour (depending on theme setting).

Example: mdi-folder-#123456

## Type of change

<!--
What type of change does your PR introduce to Homepage?
-->

- [ ] New service widget
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (please explain)

## Checklist:

- [ ] If adding a service widget or a change that requires it, I have added a corresponding PR to the [documentation](https://github.com/benphelps/homepage-docs) here: 
- [ ] If adding a new widget I have reviewed the [guidelines](https://gethomepage.dev/en/more/development/#service-widget-guidelines).
- [x] If applicable, I have checked that all tests pass with e.g. `pnpm lint`.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.

benphelps commented on src/components/resolvedicon.jsx:
```diff
@@ -33,30 +33,53 @@ export default function ResolvedIcon({ icon, width = 32, height = 32, alt = "log
+      );
+    };
 
+    if (suffix.match(`[#][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9]`) == null) {
```
Should be an else statement here, instead of checking the suffix twice.

Mega-Volti replied:
I had that first but then some check complained and it wouldn't compile properly when building the Dockerfile. This "ugly" check was the only way I got the Dockerfile to build, sadly.

benphelps replied:
Ah right, don't wrap it in an if statement at all then, the if before it guards by returning when it matches.

Mega-Volti replied:
Ah, right, I couldn't figure out why it didn't like the return-else, but I think I understand now. Fixed with the latest commit, and building the Dockerfile works now, thanks!

shamoon commented on src/components/resolvedicon.jsx:
```diff
@@ -33,13 +33,34 @@ export default function ResolvedIcon({ icon, width = 32, height = 32, alt = "log
   if (prefix in iconSetURLs) {
-    // get icon source
+    // check whether icon ends with color code
+    if (!(suffix.match(`[#][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9]`) == null)) {
+
```
```suggestion
    if (suffix.test(`[#][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9][a-f0-9]`)) {

```

use `test` to check if a regex matches

Mega-Volti replied:
Done, and I also included upper case letters since the regex was case sensitive so far.

shamoon reviewed (changes requested):
Instead of one big if / else please change this to set the appropriate const's and the output the icon, i.e. DRY

Mega-Volti commented:
> Instead of one big if / else please change this to set the appropriate const's and the output the icon, i.e. DRY

Updated. Although I had to switch from const to let. With const block scope, I had to repeat. Eliminating the repetition means I have to declare the variables outside of the if statement, hence no block scoped const, so I used let. I'm new to js, currently I don't see a better way, I'd be happy for pointers if I'm missing something.

Regarding the regex test: I updated my comment earlier and reverted the commit, the test function doesn't work on strings.

shamoon commented:
Please take a look at the last commit:

- Yea sorry that `test` snippet was written on my phone, it was incorrect as you noted
- Use regex literals
- Default vars to the current behavior and only change them if custom color detected

Let me know if you have any concerns or notice any issues

Mega-Volti commented:
Looks great, that matcher is indeed better, thanks!
No concerns or major issues from my side. I think I would have gone with an if..else statement to write the iconName and iconColor only once, instead of writing it and possibly overwiritng it again with a different value. But that's personal preference I guess. As long as it works, I'm very happy with it!

hex-developer commented:
Hello @benphelps,
Is there a chance you will implement this feature for other svg icons? I want to use some custom icons (https://tabler-icons.io) and they show up black, which doesn't work very well with my dark theme! Alternatively you could also add built-in support for tabler icons. I might submit a PR at sone point.
